### PR TITLE
Migrate ocb binary release to opentelemetry-collector-releases (see corresponding PR#608 in opentelemetry-collector-releases repo)

### LINKED ISSUE #10662: Migrate release of OpenTelemetry Collector Builder (ocb) to the opentelemetry-collector-releases repository
Currently, the ocb binary is released in the opentelemetry-collector repository, as opposed to the rest of the binaries related to the collector, including core, contrib, and k8s distributions in the opentelemetry-collector-releases repository. Since many OpenTelemetry Collector users actually build their own custom opentelemetry collector using ocb, it may be confusing to find the binary in a different repository 

I would like to migrate the release workflows from the opentelemetry-collector repository to the opentelemetry-collector-releases repository. 

Alternatively, the workflows could remain inside the opentelemetry-collector repo but be modified to publish the binaries to the opentelemetry-collector-releases repository instead of the collector repository when the release train is run. I am not opposed to either solution, but will have to get more familiar with the actions and scripts involved with the build and release process (would love to dive into any documentation we have about this process)!

I intend to add a standard dockerfile for the ocb binary, in reference to [Issue 5712](https://github.com/open-telemetry/opentelemetry-collector/issues/5712), but in that ticket and the [subsequent PR](https://github.com/open-telemetry/opentelemetry-collector/pull/6451), it was generally agreed that the ocb binary should be released to the opentelemetry-collector-releases repository prior to adding dockerfile support.

Anyone available to point me in the right direction for the build/release process for the collector? Thanks!

### PR DESCRIPTION
**Pull request by jackgopack4**
**EDIT**: After discussion with the collector SIG call this week, July 31st, I have made small tweaks to the artifact release that will not require an "extra step" of pushing tags. If these changes are accepted, the distributions will be released under tag "vX.XXX.X" and the ocb binaries will be released under "cmd/builder/vX.XXX.X" but both under this repo, opentelemetry-collector-releases. The workflow `builder-release.yaml` in releases repo will push the additional tags necessary for the builder release.

Refactoring the release process as mentioned in past issues/PRs. ocb binary is currently released in a separate release tag under this repository, which makes it harder to find and inconsistent with the rest of the binaries and images releasing in the opentelemetry-collector-releases repository.

Releasing the ocb binaries in opentelemetry-collector-releases makes this process more standardized and also opens the opportunity to release more tools alongside ocb binary down the road, like corresponding docker images (as discussed in #5712).

#### Link to tracking issue
Fixes #10662
Mentioned in [releases issue # 295](https://github.com/open-telemetry/opentelemetry-collector-releases/issues/295)
[Corresponding PR in opentelemetry-collector-releases](https://github.com/open-telemetry/opentelemetry-collector-releases/pull/608)

#### Testing
Sample binary release in [forked jackgopack4/opentelemetry-collector-releases repo](https://github.com/jackgopack4/opentelemetry-collector-releases/releases/tag/cmd%2Fbuilder%2Fv0.106.1)

#### Documentation
Updated corresponding documentation to reflect new release process

### CODE REVIEW

File: .github/workflows/sourcecode-release.yaml
Comment (jackgopack4) [inline .github/workflows/sourcecode-release.yaml:1]: the Builder release action in opentelemetry-collector used to release both the ocb binaries and the otel-collector-core source code. Now it only releases the source code, hence the file name change

File: cmd/builder/.goreleaser.yml
Comment (jackgopack4) [inline cmd/builder/.goreleaser.yml:1]: necessary to change the release repo for goreleaser

Comment (jackgopack4) [inline cmd/builder/.goreleaser.yml:1]: moved to the opentelemetry-collector-releases repository

File: docs/release.md
Comment (evan-bradley) [inline docs/release.md:80]: Is there any way we could specify the dependency in the workflows instead of requiring an additional tag? The release process is already fairly complex, it would be nice not to add any new steps.

**Review by evan-bradley — COMMENTED** [review]
Thank you for taking this on.

File: docs/release.md
Comment (jackgopack4) [reply docs/release.md:80]: I am open to suggestions and wanted to make this happen, but after looking at the release workflow (build on PR, build again on merge, then release on tag), I couldn't figure out a good way to separate the dependencies. Especially since the makefile simply downloads the prebuilt ocb release to build the different distributions. 
The other way to approach this might be using a token with cross-repo access to build the ocb release in the opentelemetry-collector repo but push it to the releases repo, but I wasn't certain if that was best practice security-wise and might make it more opaque which release is happening where.

Comment (evan-bradley) [reply docs/release.md:81]: >Especially since the makefile simply downloads the prebuilt ocb release to build the different distributions.

This is a "dumb" approach, but here's how I would do it:
* Run something like `go build  go.opentelemetry.io/collector/cmd/builder@v0.106.0 -o ./bin/ocb` to build the binary.
* Use the `cache` action to cache the binary using the tag version (and maybe OS+arch) as the cache key.
* Release OCB using the binaries from the cache (or just pull them from the build results).
* Use the cached binary to build the Collector.

We would just need to link the workflows to ensure that OCB is built before the Collector. If necessary, I'd be okay migrating our separate workflows to jobs in a "Release" workflow to allow us to just link the jobs.

>Another option (not without its downsides) is to decouple the OCB release from the Collector distros release, and use OCB v.(N-1) to build Collector v.N.

I'd be okay with this as a backup if we can't do something like the above. It would be nice to continue to build with the latest OCB to let us take advantage of Collector API updates without waiting for an additional release if we can make it work.

Comment (jackgopack4) [reply docs/release.md:81]: And running `go build` would only rely on the source code release that was generated in the `opentelemetry-collector` repository?

File: .github/workflows/check-goreleaser.yaml
Comment (jackgopack4) [inline .github/workflows/check-goreleaser.yaml:1]: goreleaser file is now located in opentelemetry-collector-releases and the check functionality is migrated to workflows there (specifically builder-test.yaml)

File: docs/release.md
Comment (evan-bradley) [inline docs/release.md:31]: I don't mind moving to numbering like this since it's easier to edit, but could we make this change in a separate PR? Including it in this one makes it harder to review.

File: docs/release.md
Comment (evan-bradley) [reply docs/release.md]: >And running `go build` would only rely on the source code release that was generated in the `opentelemetry-collector` repository

It should, yeah. So long as you point it at the module, it will only use whatever we publish for that module.

>I wasn't certain if this was something that was handled by goreleaser, which in that case, would need to be run prior to building the collector distributions.

I'm still not sure how to point Goreleaser at a remote module, but at a minimum we should be able to do something similar to the approach you took in the -releases PR to just build it twice.

File: docs/release.md
Comment (jackgopack4) [reply docs/release.md:31]: sure, I can do that.

Comment (jackgopack4) [reply docs/release.md]: fixed

File: docs/release.md
Comment (jackgopack4) [reply docs/release.md]: we publish the source code and corresponding go module here, and that gets pulled by the build script in releases. was able to publish a release in both forked repositories under jackgopack4/opentelemetry-collector and jackgopack4/opentelemetry-collector-releases

Comment (jackgopack4) [reply docs/release.md]: @evan-bradley any further thoughts/concerns on this PR or the releases PR? Thanks.

File: .chloggen/ocb-migration.yaml
Comment (mx-psi) [inline .chloggen/ocb-migration.yaml:18]:
```suggestion
subtext: ocb binaries will now be released under open-telemetry/opentelemetry-collector-releases tagged as "cmd/builder/vX.XXX.X"
```

**Review by mx-psi — APPROVED** [review]

File: docs/release.md
Comment (evan-bradley) [inline docs/release.md:92]: I'm open to other opinions, but I think I would like to keep everything under a single tag if possible. Right now we have all builder release notes grouped in with our main tag release, so users would grab the binary from one release but read the changelog in another, and that feels odd to me. I also feel like the advantages of having a single tag vs. separate tags are 50/50, so keeping a single tag would be simpler. 

The advantages I see are:
* Separate tags makes the ocb binary easier to find amongst the currently ~383 assets we publish. It also provides some logical separation between distro releases and other binary releases.
* A single tag means the release itself is easier to find (in my opinion) since the GH releases page feels geared toward a single stream of releases rather than concurrent releases of similar things. Also means the binaries and changelogs are together, and is a simpler setup.

**Review by evan-bradley — COMMENTED** [review]
Thanks for your continued work @jackgopack4. I'm overall good with this, though I did leave a comment about how we should approach tagging.

I think we should aim to merge the -releases PR before this one just to be safe. If we don't have both make it into the release, I think it would be better to double publish the release than to not publish one.

File: docs/release.md
Comment (evan-bradley) [reply docs/release.md]: Sorry for the delay, I was digging into it to see if there's a better way to approach building the source, since I'm also looking to add a binary to be released from -releases. I think that part is good to go, but I left a comment on the tagging I just want to make sure we're being deliberate about. I'll resolve this thread since I think the approach to building the binary is as good as we can do.

File: docs/release.md
Comment (mx-psi) [reply docs/release.md:92]: Can we solve discoverability with some links pointing users to where the artifacts are located? that way we have both the separation (both logical and in terms of the artifacts) as well as discoverability, at the cost of one more click for some people

Comment (evan-bradley) [reply docs/release.md:92]: I don't think the separation is super important since we essentially release everything as a single unit anyway, but that seems like a reasonable compromise.

**Review by evan-bradley — APPROVED** [review]

File: docs/release.md
Comment (jackgopack4) [reply docs/release.md:92]: I agree that having everything released under one tag would be ideal; I didn't dig deep enough into how the partial goreleaser releases worked for the distributions to implement it for builder, and plus wanted to get others' opinions on if keeping the separate tags was smart, especially if we are adding other things to this release down the line like telemetrygen, etc.

Comment (evan-bradley) [reply docs/release.md:92]: I'm still okay with a single tag. Everything is pretty tightly coupled, and I don't think the separate tags/releases will really be something we can leverage in the near future. Plus, adding ~8 binaries on top of the existing 383 doesn't really make a big difference, especially when a new Collector distro will presumably add 50+ new binaries anyway.

That said, I don't really feel strongly about this if we can link to it from the main release, and we can always change it later. I'll bring this up at tomorrow's SIG in case anyone has strong opinions.

**Review by songy23 — APPROVED** [review]
LGTM

cc release manager for the next release @dmitryax

**Review by dmitryax — APPROVED** [review]

**Pull request merged**
